testing/sdk_upgrade.py

```diff
@@ -201,50 +204,18 @@ def _upgrade_or_downgrade(
             package_name, service_name))
         sdk_plan.wait_for_completed_deployment(service_name, timeout_seconds)
 
-# Retry the decorated function several times in the event of exception_class. Raise the exception
-# normally if the number of retries is exceeded.
-def _retry(retries=3, exception_class=Exception):
-    def decorator(fn):
-        @functools.wraps(fn)
-        def wrapper(*args, **kwargs):
-            exception = None
-            for i in range(retries):
-                try:
-                    result = fn(*args, **kwargs)
-                except exception_class as e:
-                    exception = e
-                    continue
-
-                return result
-
-            raise exception
-        return wrapper
 
-    return decorator
-
-@_retry(exception_class=AttributeError)
+@retrying.retry(stop_max_attempt_number=5,
+                wait_fixed=1000,
+                retry_on_result=lambda result: result is None)
 def _get_pkg_version(package_name):
-    return re.search(
+    result = re.search(
         r'"version": "(\S+)"',
-        sdk_cmd.run_cli('package describe {}'.format(package_name), print_output=False)).group(1)
-
-# Default repo is the one at index=0.
-def _add_repo(repo_name, repo_url, prev_version, index, default_repo_package_name):
-    assert shakedown.add_package_repo(
-        repo_name,
-        repo_url,
-        index)
-    # Make sure the new default repo packages are available
-    _wait_for_new_default_version(prev_version, default_repo_package_name)
-
-
-def _add_last_repo(repo_name, repo_url, prev_version, default_repo_package_name):
-    assert shakedown.add_package_repo(
-        repo_name,
-        repo_url)
-    # Make sure the new default repo packages are available
-    _wait_for_new_default_version(prev_version, default_repo_package_name)
+        sdk_cmd.run_cli('package describe {}'.format(package_name), print_output=False))
```

Comment:
Since we're editing this anyway, could we split this into multiple steps? This may make handling errors a little easier. Also, is the output of `package describe` not already JSON from which the `"version"` field can be read directly without requiring regex?

Reply:
Restructured this a bit but it's still the same steps. Failure by Cosmos to handle the lookup is treated equivalent to Cosmos returning bad data.